### PR TITLE
SI-7511 Remove indirection of numeric methods

### PR DESCRIPTION
**Pull request by soc**
Methods on `Rich<AnyVal>`s like `abs`, `doubleValue`, `floatValue`,
`longValue`, `intValue`, `byteValue`, `shortValue`, `min`, `max`,
`signum` are slow because they go through multiple indirections
including typeclasses and boxing.

For instance, take `1L.abs`:

Instead of just computing and returning the result (like it is done in
`RichInt`), this is what happens:
- `RichLong` inherits `abs` by extending `ScalaNumberProxy[T]`
- `ScalaNumberProxy[T]` has an abstract
  `protected implicit def num: Numeric[T]`
- This method is implemented in `RichLong` and points to
  `scala.math.Numeric.LongIsIntegral`
- The actual method `abs` now calls `num.abs(self)`

### CODE REVIEW

**Comment by soc** [comment]
Review by @phaller.
I asked Rex Kerr whether he can comment on it (and maybe his idea of properly testing it against regressions): https://groups.google.com/d/topic/scala-internals/t2DndCtkhlw/discussion

File: src/library/scala/runtime/RichDouble.scala [hunk 1]
```diff
@@ -14,6 +14,35 @@ final class RichDouble(val self: Double) extends AnyVal with FractionalProxy[Dou
...
+    val l = self.toLong
+    l.toDouble == self || l == Long.MaxValue && self < Double.PositiveInfinity || l == Long.MinValue && self > Double.NegativeInfinity
+  }
+  override def isValidByte  = self.toByte.toDouble == self
+  override def isValidShort = self.toShort.toDouble == self
+  override def isValidChar  = self.toChar.toDouble == self
+  override def isValidInt   = self.toInt.toDouble == self
+  // override def isValidLong = { val l = self.toLong; l.toDouble == self && l != Long.MaxValue }
+  // override def isValidFloat = self.toFloat.toDouble == self
+  // override def isValidDouble = !java.lang.Double.isNaN(self)
```
Comment (xeno-by) [inline src/library/scala/runtime/RichDouble.scala:34]: Why comment out newly introduced code?

**Comment by ijuma** [comment]
Nice work removing useless indirections.

File: src/library/scala/runtime/RichDouble.scala (hunk 1 above)
Comment (soc) [inline src/library/scala/runtime/RichDouble.scala:34]: That part already existed, I just kept it.

Comment (xeno-by) [inline src/library/scala/runtime/RichDouble.scala:34]: Hmm why was it moved then?

Comment (soc) [inline src/library/scala/runtime/RichDouble.scala:34]: I took the chance to establish some ordering of the various methods in number classes.

**Comment by soc** [comment]
@ijuma: I prefer if all those calls go through `scala.math` (so people porting it to different runtimes don't have to change/implement things at dozens of different places).
Make everything in `math` `@inline`? (Not sure if it makes a difference...)

**Comment by adriaanm** [comment]
I believe @phaller may be travelling. @Ichoran, what say you?

**Comment by Ichoran** [comment]
I say if your JIT compiler can't remove a simple indirection off a static object, you've got worse issues than a little overhead in your math library.  I would use the scala.math versions absent some direct evidence that this imposes some performance penalty.  If there's no penalty, I'd agree with trying to make less work for people implementing different back ends.  Anyway, the patch LGTM with the proviso that I haven't tested it at all.

**Comment by adriaanm** [comment]
Maybe we should start generating these sources.

**Comment by Ichoran** [comment]
I reach for the "generate" button as soon as anyone, but this doesn't
seem worth it due to the attention one needs to pay to irregularities
(e.g. with Char) and conversion of small types to Int.  You'd spend
longer on your generator than to just do it, and this isn't the sort
of thing that requires much maintenance.  (More likely that the
generator code will break than this will.)

  --Rex

**Pull request merged**